### PR TITLE
[web] Fix shadows for arbitrary paths on PhysicalShape

### PR DESCRIPTION
**Pull request by ferhatb**
PhysicalShape shadows failed for non rect/oval shapes since the physical shape adds an svg clip path to render background. This PR adds an SVG asset behind inner clip area to render shadow of physical shape separately.

Related issues:
https://github.com/flutter/flutter/issues/65498
https://github.com/flutter/flutter/issues/66334
https://github.com/flutter/flutter/issues/57527

Tests updated in golden_compositing_test.dart

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.
- [x] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

### CODE REVIEW

File: lib/web_ui/lib/src/engine/html/clip.dart [hunk 1]
```diff
@@ -291,26 +289,68 @@ class PersistedPhysicalShape extends PersistedContainerSurface
...
+    /// potentially have a shadow due to material surface elevation.
+    ///
+    /// When there is no shadow we can simply clip a div with a background
+    /// color using an svg clip path.
+    ///
+    /// Otherwise we need to paint a svg element for the path and clip
+    /// contents against same path for shadow to work since box-shadow doesn't
+    /// take clip-path into account.
+    ///
+    /// Webkit has a bug when applying clip-path on an element that has
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/clip.dart:308]: Is there a link to the WebKit bug?

File: lib/web_ui/lib/src/engine/html/clip.dart [hunk 2]
```diff
@@ -291,26 +289,68 @@ class PersistedPhysicalShape extends PersistedContainerSurface
...
+    /// position: absolute and transform. To place clipping rectangle correctly
+    /// we size the inner container to cover full pathBounds instead of sizing
+    /// to clipping rect bounds (which is the case for elevation == 0.0 where
+    /// we shift outer/inner clip area instead to position clip-path).
+    final String svgClipPath = elevation == 0.0 ?
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/clip.dart:313]: nit: the `?` should go on the next line

File: lib/web_ui/lib/src/engine/html/clip.dart [hunk 3]
```diff
@@ -291,26 +289,68 @@ class PersistedPhysicalShape extends PersistedContainerSurface
...
+            offsetX: -pathBounds.left,
+            offsetY: -pathBounds.top,
+            scaleX: 1.0 / pathBounds.width,
+            scaleY: 1.0 / pathBounds.height)
+        : _pathToSvgClipPath(path,
+            offsetX: 0.0,
+            offsetY: 0.0,
+            scaleX: 1.0 / pathBounds.right,
+            scaleY: 1.0 / pathBounds.bottom);
+    /// If apply is called multiple times (without update) , remove prior
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/clip.dart:324]: nit: remove space before the comma

File: lib/web_ui/lib/src/engine/html/clip.dart [hunk 4]
```diff
@@ -291,26 +289,68 @@ class PersistedPhysicalShape extends PersistedContainerSurface
...
     _clipElement?.remove();
+    _svgElement?.remove();
     _clipElement =
         html.Element.html(svgClipPath, treeSanitizer: _NullTreeSanitizer());
     domRenderer.append(rootElement!, _clipElement!);
+    if (elevation == 0.0) {
+      DomRenderer.setElementStyle(
+          rootElement!, 'clip-path', 'url(#svgClip$_clipIdCounter)');
+      DomRenderer.setElementStyle(
+          rootElement!, '-webkit-clip-path', 'url(#svgClip$_clipIdCounter)');
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/clip.dart:335]: Setting style properties has been showing up in our profiles. Can we set either `clip-path` or `-webkit-clip-path` depending on the `browserEngine`, but not both?

File: lib/web_ui/lib/src/engine/html/clip.dart [hunk 5]
```diff
@@ -321,7 +361,27 @@ class PersistedPhysicalShape extends PersistedContainerSurface
...
     childContainer!.style
       ..left = '-${pathBounds.left}px'
-      ..top = '-${pathBounds.top}px';
+      ..top = '-${pathBounds.top}px'
+      ..width = '${pathBounds.right}px'
+      ..height = '${pathBounds.bottom}px';
+
+    final ui.Rect pathBounds2 = path.getBounds();
+    _svgElement = _pathToSvgElement(
+        path, SurfacePaintData()..color = color, '${pathBounds2.right}', '${pathBounds2.bottom}');
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/clip.dart:370]: Does `_applyColor` need to be revised now that this function also uses the `color` field?

File: lib/web_ui/lib/src/engine/html/clip.dart [hunk 6]
```diff
@@ -330,15 +390,16 @@ class PersistedPhysicalShape extends PersistedContainerSurface
     if (oldSurface.color != color) {
       _applyColor();
     }
-    if (oldSurface.elevation != elevation ||
+    if (oldSurface.path != path || oldSurface.elevation != elevation ||
```
Comment (yjbanov) [inline lib/web_ui/lib/src/engine/html/clip.dart:393]: Do we have a test that covers the update codepath? We frequently get bugs that only manifest when the DOM nodes are updated from a previous state.

File: lib/web_ui/lib/src/engine/html/clip.dart (hunk 1 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/clip.dart:308]: added link.

File: lib/web_ui/lib/src/engine/html/clip.dart (hunk 2 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/clip.dart:315]: done.

File: lib/web_ui/lib/src/engine/html/clip.dart (hunk 3 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/clip.dart:326]: done.

File: lib/web_ui/lib/src/engine/html/clip.dart (hunk 4 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/clip.dart]: done. Moved clip-path code into DomRenderer

File: lib/web_ui/lib/src/engine/html/clip.dart (hunk 5 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/clip.dart:366]: done.

File: lib/web_ui/lib/src/engine/html/clip.dart (hunk 6 above)
Comment (ferhatb) [reply lib/web_ui/lib/src/engine/html/clip.dart:386]: Done.

**Review by yjbanov — APPROVED** [review]
lgtm

**Pull request merged**